plugins/zify/zify.ml

```diff
@@ -0,0 +1,1121 @@
+(************************************************************************)
+(*         *   The Coq Proof Assistant / The Coq Development Team       *)
+(*  v      *   INRIA, CNRS and contributors - Copyright 1999-2019       *)
+(* <O___,, *       (see CREDITS file for the list of authors)           *)
+(*   \VV/  **************************************************************)
+(*    //   *    This file is distributed under the terms of the         *)
+(*         *     GNU Lesser General Public License Version 2.1          *)
+(*         *     (see LICENSE file for the text of the license)         *)
+(************************************************************************)
+
+open Constr
+open Names
+open Pp
+open Lazy
+
+(** [get_type_of] performs beta reduction ;
+    Is it ok for Retyping.get_type_of (Zpower_nat n q) to return (fun _ : nat => Z) q ? *)
+let get_type_of env evd e =
+  Tacred.cbv_beta env evd (Retyping.get_type_of env evd e)
+
+(** [unsafe_to_constr c] returns a [Constr.t] without considering an evar_map.
+    This is useful for calling Constr.hash *)
+let unsafe_to_constr = EConstr.Unsafe.to_constr
+
+let pr_constr env evd e = Printer.pr_econstr_env env evd e
+
+(** [get_arrow_typ evd t] returns [t1;.tn] such that t = t1 -> .. -> tn.ci_npar
+    (only syntactic matching)
+ *)
+let rec get_arrow_typ evd t =
+  match EConstr.kind evd t with
+  | Prod (a, p1, p2) (*when a.Context.binder_name = Names.Anonymous*) ->
+      p1 :: get_arrow_typ evd p2
+  | _ -> [t]
+
+(** [get_binary_arrow t] return t' such that t = t' -> t' -> t' *)
+let get_binary_arrow evd t =
+  let l = get_arrow_typ evd t in
+  match l with
+  | [] -> assert false
+  | [t1; t2; t3] -> Some (t1, t2, t3)
+  | _ -> None
+
+(** [get_unary_arrow t] return t' such that t = t' -> t' *)
+let get_unary_arrow evd t =
+  let l = get_arrow_typ evd t in
+  match l with [] -> assert false | [t1; t2] -> Some (t1, t2) | _ -> None
+
+(** [HConstr] is a map indexed by EConstr.t.
+    It should only be used using closed terms.
+ *)
+module HConstr = struct
+  module M = Map.Make (struct
+    type t = EConstr.t
+
+    let compare c c' =
+      Constr.compare (unsafe_to_constr c) (unsafe_to_constr c')
```

Comment:
@ppedrot is it the right way of doing this? Or should the type be `Constr.t` since all the terms are expected to be closed?

Reply:
Note that the difficulty is we want lookup to work on arbitrary terms (not necessarily closed ones).